### ISSUE TITLE
mysqld OOM on dev sites

### ISSUE THREAD

**Issue by cgolubi1** [buttonmen-dev/buttonmen#2841]
I've been seeing a lot more mysqld OOM events on dev and replay sites than i had previously.  I don't have any kind of numbers for this, so obviously it's *possible* that we always had this many, and i'm just paying more attention than i was for awhile.  But i think there's been a phase change.

**Comment by cgolubi1** [comment]
I do wonder if this is related to mod_pagespeed --- maybe it uses memory in a spiky way that these instances can't handle.  If that were the case, why would it be mysqld being killed instead of apache?  But i imagine there could be some reason that would happen.

**Comment by blackshadowshade** [comment]
Do we have any way of quantifying the OOM events objectively?

**Comment by cgolubi1** [comment]
Hmm, well, actually: there are a couple of reasons i think we're getting more OOMs lately, and some of them are due to improve logging on dev and replay sites / improved grepping of apache logs, that didn't exist before.  But one of them is when backups are failing with a "couldn't connect to mysql" type error.  And those i can count via e-mail messages from the backup cron job.  Here are the timestamps of those messages:

```
Date: Sun, 17 Jan 2021 00:19:03 +0000 (UTC)
Date: Wed,  3 Feb 2021 00:01:01 +0000 (UTC)
Date: Wed,  3 Feb 2021 00:01:01 +0000 (UTC)
Date: Thu,  4 Feb 2021 00:01:01 +0000 (UTC)
Date: Thu,  4 Feb 2021 00:01:01 +0000 (UTC)
Date: Fri,  5 Feb 2021 00:01:01 +0000 (UTC)
Date: Sat,  6 Feb 2021 00:01:01 +0000 (UTC)
Date: Sun,  7 Feb 2021 00:01:01 +0000 (UTC)
Date: Mon,  8 Feb 2021 00:01:01 +0000 (UTC)
Date: Tue,  9 Feb 2021 00:01:01 +0000 (UTC)
Date: Wed, 10 Feb 2021 00:01:01 +0000 (UTC)
Date: Thu, 11 Feb 2021 00:01:01 +0000 (UTC)
Date: Tue, 11 May 2021 00:05:57 +0000 (UTC)
Date: Fri,  2 Jul 2021 00:01:01 +0000 (UTC)
Date: Fri, 14 Jan 2022 00:01:01 +0000 (UTC)
Date: Mon, 17 Jan 2022 00:01:01 +0000 (UTC)
Date: Mon, 24 Jan 2022 00:01:01 +0000 (UTC)
Date: Wed, 26 Jan 2022 00:01:01 +0000 (UTC)
Date: Thu, 27 Jan 2022 00:01:01 +0000 (UTC)
Date: Fri, 28 Jan 2022 00:01:01 +0000 (UTC)
Date: Sat, 29 Jan 2022 00:01:01 +0000 (UTC)
Date: Sun, 30 Jan 2022 00:01:01 +0000 (UTC)
Date: Mon, 31 Jan 2022 00:01:01 +0000 (UTC)
Date: Thu,  3 Feb 2022 00:01:01 +0000 (UTC)
Date: Fri,  4 Feb 2022 00:01:01 +0000 (UTC)
Date: Tue, 15 Feb 2022 00:01:01 +0000 (UTC)
```

So that tells me the phase change happened around 1/14.  Now, that said, of those 12-13 recent OOMs, all but two of them were on the wildcard dev site.  So there's some argument to be made that part of what i'm seeing here is related to the code on the wildcard dev site.

However, the most recent OOM, on the 15th, was on the `flavour_text` dev site, and i have trouble imagining that *anything* was actively being done, expensive-game-wise, to that site at the time it OOMed.  That single data point pushed me in the direction of thinking something is wrong.

**Comment by cgolubi1** [comment]
Also note that the backup-time errors are an undercount; i've seen OOMs impact replay sites too (again, including replay sites like the one for Dan's python refactor which really shouldn't have anything to do with mysqld).  That's a case in which it's hard to know if there's a real phase change, or if there's always been this level of OOMs and i'm just keeping a closer eye on replay site health right now.  Subjectively, i think it's a phase change, but i don't have data.

**Comment by blackshadowshade** [comment]
I'm just thinking out loud here, so I'll ask some stupid questions. Feel free to pick up some or none of what I'm asking, I don't require answers.

- Are we actually doing backups of the dev sites?
- How long does a backup of a dev database (or staging) take? I know that the main site takes in the order of 4 minutes.
- How much memory do we have on our dev instances? Are we running on minimal instances?
- Would you expect this to actually impact the live site at all, considering that the dev sites have their own databases running on the same instance, but the live site has a database running on a separate instance?

Also, if we're going to have OOM errors affecting the main site, this is another good reason to have any stats running on a separate instance, so that they don't impact performance on the main site.

**Comment by cgolubi1** [comment]
Real quick answer: i'm confident none of this is of direct concern to prod or staging, because on those sites, the database runs on RDS, not on the instance itself.  So whatever the DBs are doing, they're not in-scope for the OOM-killer in the instance OS.

More later.

**Comment by jl8e** [comment]
Is it possible to check when there was last player activity on the dev sites?

My guess as to why wildcard is blowing up so often is that WC is a big die, and players are making custom buttons with a lot of them, and it's the hit-table problem again.

**Comment by blackshadowshade** [comment]
While this might be true, it doesn't explain the OOM problem on the flavour text dev site.

**Comment by blackshadowshade** [comment]
And yes, there's a way to check when there was last players logged on, via the Who's online page.

**Comment by jl8e** [comment]
> While this might be true, it doesn't explain the OOM problem on the flavour text dev site.

True. It should be easy enough to check whether there's ever been a game played on that one.

**Comment by blackshadowshade** [comment]
I took a look just now at fleeting.dev (which is where we had the flavour text updated), and only two people have logged into it (apart from me just now):

dvstest2 | 2 days
glassonion | 3 days

Based on that, there's been no game playing happening.

**Comment by blackshadowshade** [comment]
Just checked the history page to double check, and the last move played by dvstest2 is on 6 Jan, so obviously on staging.
